### PR TITLE
Add compilation start / stop ETW events

### PR DESCRIPTION
**Pull request by jaredpar**
This adds a compilation start / stop event pair to ETW for the compiler server. When looking at a profile of VBCSCompiler it allows us to isolate out the time range for specific projects. This is very helpful when tracking down compiler perf issues, particularly for issues that only manifest when a project is built in the context of a full build vs. compiling independently.

### CODE REVIEW

**Comment by jaredpar** [comment]
@dotnet/roslyn-compiler, @chsienki PTAL

File: src/Compilers/Core/MSBuildTask/ManagedCompiler.cs [hunk 1]
```diff
@@ -574,6 +586,20 @@ internal int ExecuteTool(string pathToTool, string responseFileCommands, string
             }
 
             return ExitCode;
+
+            // Construct the friendly name for the compilation. This does not need to be unique. Instead
+            // it's used by developers to understand what compilation is running on the server.
+            string GetRequestId()
```
Comment (chsienki) [inline src/Compilers/Core/MSBuildTask/ManagedCompiler.cs:592]: Love this

File: src/Compilers/Core/Portable/CodeAnalysisEventSource.cs [hunk 2]
```diff
@@ -95,6 +96,12 @@ internal unsafe void NodeTransform(int nodeHashCode, string name, string tableTy
             }
         }
 
+        [Event(7, Message = "Compilation {0} started", Keywords = Keywords.Performance, Level = EventLevel.Informational, Opcode = EventOpcode.Start, Task = Tasks.Compilation)]
```
Comment (chsienki) [inline src/Compilers/Core/Portable/CodeAnalysisEventSource.cs:99]: Do we want to disambiguate that this is a CompilerServer Compilation?

File: src/Compilers/Core/MSBuildTask/Microsoft.CSharp.Core.targets [hunk 3]
```diff
@@ -134,6 +134,7 @@
          Platform="$(PlatformTarget)"
          Prefer32Bit="$(Prefer32Bit)"
          PreferredUILang="$(PreferredUILang)"
+         ProjectName="$(MSBuildProjectName)"
```
Comment (chsienki) [inline src/Compilers/Core/MSBuildTask/Microsoft.CSharp.Core.targets:137]: Do we still have copies of this task floating around in other repos? I thought WPF had one but I couldn't find it.

**Review by chsienki — APPROVED** [review]
LGTM

File: src/Compilers/Core/MSBuildTask/ManagedCompiler.cs [hunk 4]
```diff
@@ -574,6 +586,20 @@ internal int ExecuteTool(string pathToTool, string responseFileCommands, string
             }
 
             return ExitCode;
+
+            // Construct the friendly name for the compilation. This does not need to be unique. Instead
+            // it's used by developers to understand what compilation is running on the server.
```
Comment (jjonescz) [inline src/Compilers/Core/MSBuildTask/ManagedCompiler.cs:591]: Can it happen in reality that developers want to disambiguate requests? Maybe the friendly name could also include a GUID for that.

Also someone could be using the Csc task without setting ProjectName and TargetFramework, I imagine, and debugging the logs would be easier if there was a fallback GUID.

File: src/Compilers/Core/MSBuildTask/ManagedCompiler.cs (hunk 1 above)
Comment (jjonescz) [inline src/Compilers/Core/MSBuildTask/ManagedCompiler.cs:592]: Shouldn't local functions start with lowercase letter?
```suggestion
            string getRequestId()
```

File: src/Compilers/Core/MSBuildTask/ManagedCompiler.cs [hunk 5]
```diff
@@ -574,6 +586,20 @@ internal int ExecuteTool(string pathToTool, string responseFileCommands, string
...
+            string GetRequestId()
+            {
+                if (!string.IsNullOrEmpty(ProjectName))
+                {
+                    return string.IsNullOrEmpty(TargetFramework)
+                        ? ProjectName
+                        : $"{ProjectName} ({TargetFramework})";
+                }
+
+                return $"Unnamed compilation";
```
Comment (jjonescz) [inline src/Compilers/Core/MSBuildTask/ManagedCompiler.cs:601]:
```suggestion
                return "Unnamed compilation";
```

File: src/Compilers/Core/MSBuildTask/ManagedCompiler.cs (hunk 4 above)
Comment (jaredpar) [reply src/Compilers/Core/MSBuildTask/ManagedCompiler.cs:591]: Yep. Will add the `Guid` in the fallback case.

File: src/Compilers/Core/MSBuildTask/ManagedCompiler.cs (hunk 1 above)
Comment (jaredpar) [reply src/Compilers/Core/MSBuildTask/ManagedCompiler.cs:592]: Yep

File: src/Compilers/Core/MSBuildTask/ManagedCompiler.cs (hunk 5 above)
Comment (jaredpar) [reply src/Compilers/Core/MSBuildTask/ManagedCompiler.cs:601]: Had to add `$` when I put the GUID back in.

File: src/Compilers/Core/MSBuildTask/Microsoft.CSharp.Core.targets (hunk 3 above)
Comment (jaredpar) [reply src/Compilers/Core/MSBuildTask/Microsoft.CSharp.Core.targets:137]: I _think_ we got rid of the other places where `Csc` task was invoked directly. There have been a few other cases where we added properties recently and haven't had much issue. If we find some though can update them.

File: src/Compilers/Core/Portable/CodeAnalysisEventSource.cs (hunk 2 above)
Comment (jaredpar) [reply src/Compilers/Core/Portable/CodeAnalysisEventSource.cs:99]: Not sure if you're suggesting the message should be diff or I should use different names for the methods here.

Comment (chsienki) [reply src/Compilers/Core/Portable/CodeAnalysisEventSource.cs:99]: I would probably use different event names, just so its super obvious from an ETW trace what it represents. We don't have it for non-server compilations right now, but there seems like a future where we might?

**Review by jjonescz — APPROVED** [review]

File: src/Compilers/Core/MSBuildTask/Microsoft.CSharp.Core.targets (hunk 3 above)
Comment (jaredpar) [reply src/Compilers/Core/MSBuildTask/Microsoft.CSharp.Core.targets:137]: I was wrong https://github.com/dotnet/msbuild/pull/9786

**Pull request merged**